### PR TITLE
Feat/eagle support

### PR DESCRIPTION
**Pull request by pengchengneo**
https://github.com/sgl-project/sglang-jax/issues/192
complete Eagle Worker main process

`python3 -u -m sgl_jax.launch_server   --model-path Qwen/Qwen3-32B   --trust-remote-code   --device=tpu    --random-seed=3   --mem-fraction-static=0.8   --max-prefill-tokens=4096   --max-running-requests=16  --attention-backend=fa   --dtype=bfloat16  --port 30000 --host 0.0.0.0 --disable-overlap-schedule --tp-size 4 --speculative-algorithm EAGLE3 --speculative-draft-model-path  AngelSlim/Qwen3-32B_eagle3 --page-size 64 --speculative-eagle-topk 1 --speculative-num-steps 3 --speculative-num-draft-tokens 4 --speculative-draft-model-revision 67caf31f9062d7ab64872e0a111d499bc16cd205`

image

### CODE REVIEW

File: python/sgl_jax/srt/mem_cache/chunk_cache.py [hunk 1]
```diff
@@ -39,6 +39,9 @@ def cache_finished_req(self, req: Req):
             # For decode server: if req.output_ids is empty, we want to free all req.origin_input_ids
             : len(req.origin_input_ids) + max(len(req.output_ids) - 1, 0),
         ]
+        print(
```
Comment (jimoosciuc) [inline python/sgl_jax/srt/mem_cache/chunk_cache.py]: debug log?

File: python/sgl_jax/srt/utils/weight_utils.py [hunk 2]
```diff
@@ -280,9 +286,10 @@ def _handle_single_weight(
 
         if mapping.reshape is not None:
             processed_weight = jnp.reshape(processed_weight, mapping.reshape)
-
         if mapping.head_dim_padding and self.head_dim_pad > 0:
```
Comment (jimoosciuc) [inline python/sgl_jax/srt/utils/weight_utils.py]: remove？

File: python/sgl_jax/srt/utils/weight_utils.py [hunk 3]
```diff
@@ -77,9 +78,9 @@ def __init__(
             model_config, "head_dim", self.hidden_size // self.num_heads
         )
 
-        self.head_dim = (self.head_dim_original + 127) // 128 * 128
-        self.head_dim_pad = self.head_dim - self.head_dim_original
-
+        # self.head_dim = (self.head_dim_original + 127) // 128 * 128
+        self.head_dim_pad = (self.head_dim_original + 127) // 128 * 128 - self.head_dim_original
```
Comment (jimoosciuc) [inline python/sgl_jax/srt/utils/weight_utils.py:82]: remove？

File: python/sgl_jax/srt/speculative/pallas/tree_speculative_sampling_target_only_kernel.py [hunk 4]
```diff
@@ -0,0 +1,555 @@
...
+        #     return new_sum, new_sum
+        # final_sum, cumulative_sums = scan(
+        #     f=cumulative_sum_step,
+        #     init=jnp.array(0),
+        #     xs=arr
+        # )
+        # return cumulative_sums
+
+    def _sampling_from_prob(probs, threshold):
+        print(f"{probs.shape=} {threshold.shape=}")
```
Comment (jimoosciuc) [inline python/sgl_jax/srt/speculative/pallas/tree_speculative_sampling_target_only_kernel.py]: debug log?

File: python/sgl_jax/srt/models/llama_eagle3.py [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+
+    def _create_llama_ealge3_weight_mappings(self):
+        # mappings = super()._create_llama_weight_mappings()
+        mappings = {}
+        mappings["d2t"] = WeightMapping(
+            target_path="hot_token_ids",
+            sharding=(None,),
+            transpose=False,
+        )
+        # mappings["t2d"] = WeightMapping(
```
Comment (jimoosciuc) [inline python/sgl_jax/srt/models/llama_eagle3.py]: remove?

File: python/sgl_jax/srt/kernels/ragged_paged_attention/ragged_paged_attention.py [hunk 6]
```diff
@@ -136,9 +153,21 @@ def ref_ragged_paged_attention(
...
-        kv_span = jax.lax.broadcasted_iota(jnp.int32, attn.shape, 2)
-        mask = q_span < kv_span
+        if causal:
+            q_span = (kv_len - q_len) + jax.lax.broadcasted_iota(jnp.int32, attn.shape, 1)
+            kv_span = jax.lax.broadcasted_iota(jnp.int32, attn.shape, 2)
+            mask = q_span < kv_span
+        else:
+            mask_start = cu_mask_lens[i]
+            mask_end = cu_mask_lens[i + 1]
+            print(f"mask_start,  mask_end {mask_start} {mask_end}")
```
Comment (jimoosciuc) [inline python/sgl_jax/srt/kernels/ragged_paged_attention/ragged_paged_attention.py]: remove?

File: python/sgl_jax/srt/mem_cache/chunk_cache.py (hunk 1 above)
Comment (pengchengneo) [reply python/sgl_jax/srt/mem_cache/chunk_cache.py]: fixed

File: python/sgl_jax/srt/utils/weight_utils.py (hunk 2 above)
Comment (pengchengneo) [reply python/sgl_jax/srt/utils/weight_utils.py]: I will refactor this part in issue https://github.com/sgl-project/sglang-jax/issues/389

File: python/sgl_jax/srt/utils/weight_utils.py (hunk 3 above)
Comment (pengchengneo) [reply python/sgl_jax/srt/utils/weight_utils.py:82]: fixed

File: python/sgl_jax/srt/speculative/pallas/tree_speculative_sampling_target_only_kernel.py (hunk 4 above)
Comment (pengchengneo) [reply python/sgl_jax/srt/speculative/pallas/tree_speculative_sampling_target_only_kernel.py]: fixed

File: python/sgl_jax/srt/models/llama_eagle3.py (hunk 5 above)
Comment (pengchengneo) [reply python/sgl_jax/srt/models/llama_eagle3.py]: fixed

File: python/sgl_jax/srt/kernels/ragged_paged_attention/ragged_paged_attention.py (hunk 6 above)
Comment (pengchengneo) [reply python/sgl_jax/srt/kernels/ragged_paged_attention/ragged_paged_attention.py]: fixed

**Comment by SiqiLi-Fighting** [comment]
> Have you tested the availability of multi-machine, multiprocessor speculative decoding?

no test for multi host yet, I will test myself to check multi machine compatibility laterly and paste command here

**Comment by SiqiLi-Fighting** [comment]
> > Have you tested the availability of multi-machine, multiprocessor speculative decoding?
> 
> no test for multi host yet, I will test myself to check multi machine compatibility laterly and paste command here

image

` python -u -m sgl_jax.launch_server   --model-path Qwen/Qwen3-32B   --trust-remote-code   --device=tpu    --mem-fraction-static=0.8   --max-prefill-tokens=4096   --max-running-requests=16  --decode-log-interval=1   --enable-request-time-stats-logging   --attention-backend=fa   --dtype=bfloat16  --port 30000 --host 0.0.0.0 --disable-overlap-schedule --tp-size 4 --speculative-algorithm EAGLE3 --speculative-draft-model-path  AngelSlim/Qwen3-32B_eagle3 --page-size 64 --speculative-eagle-topk 1 --speculative-num-steps 3 --speculative-num-draft-tokens 4 --speculative-draft-model-revision 67caf31f9062d7ab64872e0a111d499bc16cd205 --nnodes 4  --dist-init-addr=10.138.0.36:10011 --tp-size 16 --node-rank 0 --disable-precompile`

it is compatibile with multi-host

File: python/sgl_jax/srt/kernels/speculative/kernel.py [hunk 7]
```diff
@@ -0,0 +1,349 @@
+import jax
+import jax.numpy as jnp
```
Comment (Iamleos) [inline python/sgl_jax/srt/kernels/speculative/kernel.py:2]: 1. Move all pallas kernels to folder ```python/sgl_jax/srt/kernels```
2. Add performance test for all kernels. Refer to ```benchmark/kernels/update_kv_cache/bench_update_kv_cache.py```
3. Add it to PR Test ci workflow. Refer to test suites ```kernel-performance-test-tpu-v6e-1``` defined in test/srt/run_suite.py.

File: test/srt/test_eagle_tree_build.py [hunk 8]
```diff
@@ -0,0 +1,834 @@
+#!/usr/bin/env python3
+
+import jax
+import jax.numpy as jnp
+
+from sgl_jax.srt.speculative.eagle_util import (
```
Comment (Iamleos) [inline test/srt/test_eagle_tree_build.py]: 1. Use ```CustomTestCase``` to wrap unit test cases.
2. Move this unit test to folder ```python/sgl_jax/test```
3. Add it PR Test ci workflow. Refer to test suites ```unit-test-tpu-v6e-1``` defined in ```test/srt/run_suite.py```.

File: test/srt/test_speculative_decoding.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+            base_url=self.base_url,
+            model=self.model,
+            eval_name="mmlu",
+            num_examples=512,
+            num_threads=64,
+            max_tokens=1024,
+        )
+
+        metrics = run_eval(args)
+        self.assertGreater(metrics["score"], 0.45)
```
Comment (Iamleos) [inline test/srt/test_speculative_decoding.py:86]: 1. 512 num_examples are too many. 300 is enough.
2. 0.45 baseline maybe too low
3. Add it to PR Test ci workflow. Refer to test suites ```e2e-test-tpu-v6e-1``` or ```e2e-test-tpu-v6e-4``` defined in ```test/srt/run_suite.py```.

Comment (pengchengneo) [reply test/srt/test_speculative_decoding.py:86]: I will do this after performance-optimazation

File: test/srt/test_eagle_tree_build.py (hunk 8 above)
Comment (pengchengneo) [reply test/srt/test_eagle_tree_build.py]: fixed

File: python/sgl_jax/srt/kernels/speculative/kernel.py (hunk 7 above)
Comment (pengchengneo) [reply python/sgl_jax/srt/kernels/speculative/kernel.py:2]: fixed

**Review by Iamleos — APPROVED** [review]
LGTM
